File: sgf-poker/sgf-poker-ios/Core/UseCases/UpdatePlayerUseCase.swift
// Core/UseCases/UpdatePlayerUseCase.swift

import Foundation

/// Applies a mutation to an existing player and persists the result.
/// Used for toggling membership, founder status, and updating points.
final class UpdatePlayerUseCase {

    private let repository: PlayerRepository

    init(repository: PlayerRepository) {
        self.repository = repository
    }

    /// Fetches the player, applies the mutation closure, then saves.
    /// - Parameters:
    ///   - id: The UUID of the player to update.
    ///   - mutation: A closure that receives an `inout Player` to mutate.
    /// - Throws: `PlayerError.playerNotFound` if the player does not exist.
    @discardableResult
    func execute(id: UUID, mutation: (inout Player) -> Void) throws -> Player {
        var players = try repository.fetchAll()

        guard let index = players.firstIndex(where: { $0.id == id }) else {
            throw PlayerError.playerNotFound(id: id.uuidString)
        }

        mutation(&players[index])
        try repository.save(players[index])
        return players[index]
    }
}

File: sgf-poker-android/app/src/main/java/com/sgf/poker/usecases/player/UpdatePlayerUseCase.java
package com.sgf.poker.usecases.player;

import com.sgf.poker.data.repository.PlayerRepository;
import com.sgf.poker.domain.error.PlayerException;
import com.sgf.poker.domain.model.Player;

import java.util.function.UnaryOperator;

/** Applies a transformation to a player and persists the result. */
public class UpdatePlayerUseCase {

    private final PlayerRepository repository;

    public UpdatePlayerUseCase(PlayerRepository repository) {
        this.repository = repository;
    }

    public Player execute(String playerId, UnaryOperator<Player> mutation) {
        var players = repository.fetchAll();
        var existing = players.stream()
                .filter(p -> p.getId().equals(playerId))
                .findFirst()
                .orElseThrow(() -> PlayerException.notFound(playerId));

        var updated = mutation.apply(existing);
        repository.save(updated);
        return updated;
    }
}
